File: WFM-Planning-Suite/src/wfm_io/files.py
"""File import/export and profile persistence."""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path
from typing import Any

import pandas as pd

from core.datetime.parser import parse_series, resolve_date_format
from core.models import Profile, RawUpload, ValidationIssue

EXPECTED_COLUMNS = {
    "timestamp": ["timestamp", "datetime", "date", "time", "interval"],
    "volume": ["volume", "calls", "contacts", "interactions", "count"],
    "aht": ["aht", "handle_time", "avg_handle_time", "talk_time"],
}


def _match_column(headers: list[str], aliases: list[str]) -> str | None:
    lower = {h.lower().strip(): h for h in headers}
    for alias in aliases:
        if alias in lower:
            return lower[alias]
    return None


def auto_map_columns(headers: list[str]) -> dict[str, str]:
    mapping: dict[str, str] = {}
    for canonical, aliases in EXPECTED_COLUMNS.items():
        matched = _match_column(headers, aliases)
        if matched:
            mapping[canonical] = matched
    return mapping


def load_upload(file_path: str | Path, column_mapping: dict[str, str] | None = None) -> RawUpload:
    path = Path(file_path)
    if path.suffix.lower() in (".xlsx", ".xls"):
        df = pd.read_excel(path)
    else:
        df = pd.read_csv(path)

    headers = list(df.columns)
    mapping = column_mapping or auto_map_columns(headers)
    issues: list[ValidationIssue] = []

    if "timestamp" not in mapping:
        issues.append(ValidationIssue("error", "MISSING_TIMESTAMP", "No timestamp column found or mapped"))
    if "volume" not in mapping:
        issues.append(ValidationIssue("error", "MISSING_VOLUME", "No volume column found or mapped"))

    rows: list[dict[str, Any]] = []
    date_values = df[mapping["timestamp"]].astype(str).tolist() if "timestamp" in mapping else []

    order, fmt = resolve_date_format(date_values) if date_values else (None, "")
    timestamps: list[datetime] = []
    if date_values and order is not None:
        timestamps, order, fmt = parse_series(date_values, order)

    seen: set[datetime] = set()
    for i, record in enumerate(df.to_dict(orient="records")):
        row: dict[str, Any] = {}
        if "timestamp" in mapping and i < len(timestamps):
            row["timestamp"] = timestamps[i]
            if row["timestamp"] in seen:
                issues.append(ValidationIssue("warning", "DUPLICATE_TIMESTAMP", f"Duplicate timestamp at row {i}", i))
            seen.add(row["timestamp"])
        if "volume" in mapping:
            vol = float(record.get(mapping["volume"], 0) or 0)
            row["volume"] = vol
            if vol < 0:
                issues.append(ValidationIssue("error", "NEGATIVE_VOLUME", f"Negative volume at row {i}", i))
        if "aht" in mapping:
            row["aht"] = float(record.get(mapping["aht"], 0) or 0)
        rows.append(row)

    date_range = None
    if timestamps:
        date_range = (min(timestamps), max(timestamps))

    return RawUpload(
        rows=rows,
        date_format=fmt,
        granularity_minutes=30,
        column_mapping=mapping,
        issues=issues,
        date_range=date_range,
    )


def save_profile(profile: Profile, path: str | Path) -> None:
    Path(path).write_text(json.dumps(profile.to_dict(), indent=2))


def load_profile(path: str | Path) -> Profile:
    data = json.loads(Path(path).read_text())
    return Profile.from_dict(data)


def export_report_excel(report_data: dict[str, Any], path: str | Path) -> None:
    with pd.ExcelWriter(path, engine="openpyxl") as writer:
        for sheet_name, rows in report_data.get("sheets", {}).items():
            pd.DataFrame(rows).to_excel(writer, sheet_name=sheet_name[:31], index=False)

File: WFM-Planning-Suite/web/src/core/io/files.ts
import Papa from 'papaparse'
import * as XLSX from 'xlsx'
import { parseSeries } from '../datetime/parser'
import type { DataRow, RawUpload, ValidationIssue } from '../models'

const COLUMN_ALIASES: Record<string, string[]> = {
  timestamp: ['timestamp', 'datetime', 'date', 'time', 'interval'],
  volume: ['volume', 'calls', 'contacts', 'interactions', 'count'],
  aht: ['aht', 'handle_time', 'avg_handle_time', 'talk_time'],
}

function matchColumn(headers: string[], aliases: string[]): string | null {
  const lower = Object.fromEntries(headers.map((h) => [h.toLowerCase().trim(), h]))
  for (const a of aliases) if (lower[a]) return lower[a]
  return null
}

function autoMap(headers: string[]): Record<string, string> {
  const mapping: Record<string, string> = {}
  for (const [canonical, aliases] of Object.entries(COLUMN_ALIASES)) {
    const m = matchColumn(headers, aliases)
    if (m) mapping[canonical] = m
  }
  return mapping
}

export function parseUploadFile(file: File): Promise<RawUpload> {
  return new Promise((resolve, reject) => {
    const ext = file.name.split('.').pop()?.toLowerCase()
    const reader = new FileReader()

    reader.onload = () => {
      try {
        let records: Record<string, string>[] = []
        if (ext === 'xlsx' || ext === 'xls') {
          const wb = XLSX.read(reader.result, { type: 'array' })
          const sheet = wb.Sheets[wb.SheetNames[0]]
          records = XLSX.utils.sheet_to_json(sheet, { defval: '' }) as Record<string, string>[]
        } else {
          const parsed = Papa.parse<Record<string, string>>(reader.result as string, { header: true, skipEmptyLines: true })
          records = parsed.data
        }

        const headers = records.length ? Object.keys(records[0]) : []
        const mapping = autoMap(headers)
        const issues: ValidationIssue[] = []
        if (!mapping.timestamp) issues.push({ severity: 'error', code: 'MISSING_TIMESTAMP', message: 'No timestamp column found' })
        if (!mapping.volume) issues.push({ severity: 'error', code: 'MISSING_VOLUME', message: 'No volume column found' })

        const dateValues = records.map((r) => String(r[mapping.timestamp] ?? ''))
        let timestamps: Date[] = []
        let dateFormat = ''
        if (mapping.timestamp && dateValues.some((v) => v.trim())) {
          const [parsed, , fmt] = parseSeries(dateValues)
          timestamps = parsed
          dateFormat = fmt
        }

        const seen = new Set<number>()
        const rows: DataRow[] = records.map((rec, i) => {
          const row: DataRow = { timestamp: timestamps[i] ?? new Date(), volume: 0 }
          if (mapping.volume) {
            row.volume = parseFloat(String(rec[mapping.volume])) || 0
            if (row.volume < 0) issues.push({ severity: 'error', code: 'NEGATIVE_VOLUME', message: `Negative volume at row ${i}`, rowIndex: i })
          }
          if (mapping.aht) row.aht = parseFloat(String(rec[mapping.aht])) || 0
          const ts = row.timestamp.getTime()
          if (seen.has(ts)) issues.push({ severity: 'warning', code: 'DUPLICATE', message: `Duplicate at row ${i}`, rowIndex: i })
          seen.add(ts)
          return row
        })

        const dateRange: [Date, Date] | null = timestamps.length
          ? [new Date(Math.min(...timestamps.map((t) => t.getTime()))), new Date(Math.max(...timestamps.map((t) => t.getTime())))]
          : null

        resolve({ rows, dateFormat, issues, dateRange })
      } catch (e) {
        reject(e)
      }
    }
    reader.onerror = () => reject(reader.error)

    if (ext === 'xlsx' || ext === 'xls') reader.readAsArrayBuffer(file)
    else reader.readAsText(file)
  })
}

export function exportReportExcel(sheets: Record<string, object[]>, filename: string) {
  const wb = XLSX.utils.book_new()
  for (const [name, data] of Object.entries(sheets)) {
    XLSX.utils.book_append_sheet(wb, XLSX.utils.json_to_sheet(data), name.slice(0, 31))
  }
  XLSX.writeFile(wb, filename)
}
